clusterDeployer.py

```diff
@@ -33,6 +34,149 @@ def match_to_proper_version_format(version_cluster_config: str) -> str:
     return match.group(0)
 
 
+def _update_etc_resolvconf(setup: bool = True) -> None:
+    rc = "/etc/resolv.conf"
+    rc_orig = "/etc/resolv.conf.cda-orig"
+    rc_local = "/etc/resolv.conf.cda-local"
+
+    lh = host.LocalHost()
+
+    # The user can disable touching /etc/resolv.conf by
+    #
+    #   $ sudo touch /etc/.resolv.conf.cda-dont-touch
+    #
+    # We will still write our "/etc/resolv.conf.cda*" files, but
+    # it's up to the user to redirect symlinks accordingly.
+    dont_touch = os.path.exists("/etc/.resolv.conf.cda-dont-touch")
+
+    if not setup:
+        if dont_touch:
+            return
+        if not os.path.exists(rc_orig):
+            return
+        try:
+            lnk = os.readlink(rc)
+        except Exception:
+            return
+        if not os.path.basename(lnk).startswith("resolv.conf.cda."):
+            # We only restore, if /etc/resolv.conf is a symlink to "resolv.conf-cda.*"
+            # which indicates that the file was created by us.
+            return
+        dst = os.path.basename(rc_orig)
+        if lnk == dst:
+            return
+        logger.info(f"resolv.conf: restore /etc/resolv.conf to link to original {rc_orig}")
+        lh.run_or_die(shlex.join(["ln", "-snf", dst, rc]))
+        return
+
+    content_rc = b""
+    if not os.path.exists(rc_orig):
+        try:
+            with open(rc, "rb") as f:
+                content_rc = f.read()
+        except IOError:
+            pass
+        # Backup the original resolv.conf file to "rc_orig"
+        if content_rc == b"":
+            pass
+        elif b"Written by cluster-deployment-automation" in content_rc:
+            content_rc = b""
+        elif b"Generated by NetworkManager" in content_rc and os.path.exists("/run/NetworkManager/resolv.conf"):
+            # This file seems to be written by NetworkManager. Symlink there.
+            logger.info(f"resolv.conf: point {rc_orig} to /run/NetworkManager/resolv.conf")
+            lh.run_or_die(shlex.join(["ln", "-snf", "/run/NetworkManager/resolv.conf", rc_orig]))
```

Comment:
why are we shlex.join'ing outside instead of caliing

lh.run_or_die(f"ln -snf {nm_resolve_path} {rc_orig}")

Reply:
what is better here? Call shlex at caller or callee site?

Reply:
IMO, if we are (very) sure that strings contain no odd characters, then a plain `host.run(f"program {arg1} {arg2}")` is fine.

If we are not entirely confident, and if the argument is not a shell script, then (with `th/host-run` branch), the preferred way would be `host.run(["program", arg1, arg2])` which lets the callee automatically quote the values.

Finally, if we are not entirely confident, and the argument is an actual script, then the caller needs to ensure that the shell script is valid. For example, `host.run(f"program {shlex.quote(arg1)} | grep foo")`.